File: app/src/main/java/com/lazysoul/kotlinwithandroid/ui/main/MainMvpPresenterImpl.java
package com.lazysoul.kotlinwithandroid.ui.main;

import com.lazysoul.kotlinwithandroid.common.BaseMvpView;
import com.lazysoul.kotlinwithandroid.common.RxPresenter;
import com.lazysoul.kotlinwithandroid.datas.Todo;
import com.lazysoul.kotlinwithandroid.singletons.TodoManager;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.Consumer;
import io.reactivex.subjects.PublishSubject;

/**
 * Created by Lazysoul on 2017. 7. 9..
 */

class MainMvpPresenterImpl<MvpView extends BaseMvpView> extends RxPresenter implements MainMvpPresenter<MvpView> {

    private MainMvpView view;

    private PublishSubject<String> searchTextChangeSubject = PublishSubject.create();

    MainMvpPresenterImpl() {
        add(searchTextChangeSubject
            .throttleLast(1, TimeUnit.SECONDS)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(new Consumer<String>() {
                @Override
                public void accept(String text) throws Exception {
                    ArrayList<Todo> searchedList = TodoManager.search(text);
                    if (searchedList.isEmpty()) {
                        view.showEmtpyView();
                    } else {
                        view.onUpdateTodoList(searchedList);
                    }
                }
            }));
    }

    @Override
    public void attachView(MvpView view) {
        this.view = (MainMvpView) view;
    }

    @Override
    public void destroy() {
        dispose();
    }

    @Override
    public void createTodoSamples() {
        view.onCreatedSampes(TodoManager.createSamples());
    }

    @Override
    public void loadTodoList() {
        List<Todo> todoList = TodoManager.getTodoList();
        if (null != todoList && todoList.isEmpty()) {
            view.showEmtpyView();
        } else {
            view.onUpdateTodoList(todoList);
        }
    }

    @Override
    public void insert(int id, String body) {
        view.onCreatedTodo(TodoManager.insert(id, body));
    }

    @Override
    public void update(int id, String body) {
        Todo todo = TodoManager.update(id, body);
        if (todo != null) {
            view.onUpdateTodo(todo);
        }
    }

    @Override
    public void checked(int id, boolean isChecked) {
        TodoManager.checked(id, isChecked);
    }

    @Override
    public void searchQuery(String text) {
        searchTextChangeSubject.onNext(text);
    }

    @Override
    public void searchFinish() {
        loadTodoList();
    }
}

File: app/src/main/kotlin/com/lazysoul/kotlinwithandroid/ui/main/MainMvpPresenterImpl.kt
package com.lazysoul.kotlinwithandroid.ui.main

import com.lazysoul.kotlinwithandroid.common.BaseMvpView
import com.lazysoul.kotlinwithandroid.common.RxPresenter
import com.lazysoul.kotlinwithandroid.singletons.TodoManager
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.PublishSubject
import java.util.concurrent.TimeUnit

/**
 * Created by Lazysoul on 2017. 7. 9..
 */

class MainMvpPresenterImpl<in MvpView : BaseMvpView> : RxPresenter(), MainMvpPresenter<MvpView> {

    private lateinit var view: MainMvpView

    private val searchTextChangeSubject = PublishSubject.create<String>()

    init {
        add(searchTextChangeSubject
            .throttleLast(1, TimeUnit.SECONDS)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe { text ->
                val searchedList = TodoManager.search(text)
                if (searchedList.isEmpty()) {
                    view.showEmtpyView()
                } else {
                    view.onUpdateTodoList(searchedList)
                }
            })
    }

    override fun attachView(view: MvpView) {
        this.view = view as MainMvpView
    }

    override fun destroy() {
        dispose()
    }

    override fun createTodoSamples() {
        view.onCreatedSampes(TodoManager.createSamples())
    }

    override fun loadTodoList() {
        val todoList = TodoManager.getTodoList()
        if (todoList.isEmpty()) {
            view.showEmtpyView()
        } else {
            view.onUpdateTodoList(todoList)
        }
    }

    override fun insert(id: Int, body: String) {
        view.onCreatedTodo(TodoManager.insert(id, body))
    }

    override fun update(id: Int, body: String) {
        val todo = TodoManager.update(id, body)
        if (todo != null) {
            view.onUpdateTodo(todo)
        }
    }

    override fun checked(id: Int, isChecked: Boolean) {
        TodoManager.checked(id, isChecked)
    }

    override fun searchQuery(text: String) {
        searchTextChangeSubject.onNext(text)
    }

    override fun searchFinish() {
        loadTodoList()
    }
}
